src/Middleware/HeaderPropagation/src/HeaderPropagationMiddleware.cs

```diff
@@ -33,7 +33,8 @@ public HeaderPropagationMiddleware(RequestDelegate next, IOptions<HeaderPropagat
             _values = values ?? throw new ArgumentNullException(nameof(values));
         }
 
-        public Task Invoke(HttpContext context)
+        // This needs to be async as otherwise the AsyncLocal could bleed across requests, see https://github.com/aspnet/AspNetCore/issues/13991.
+        public async Task Invoke(HttpContext context)
         {
             // We need to intialize the headers because the message handler will use this to detect misconfiguration.
             var headers = _values.Headers ??= new Dictionary<string, StringValues>(StringComparer.OrdinalIgnoreCase);
```

Comment:
Can we make this new up one every time as well?

Reply:
We could but it would be a breaking change. Any particular reason?

Reply:
I'm not sure why there's even a null check here? Who else would be initializing Headers before this component? If you had multiple instances of the middleware?

Reply:
Right. We should just change it to create a new one every time. I'm not sure why this is being stored in the first place (that's what caused the original bug).

Reply:
HeaderPropagationValues is the async local that's consumed later, so it does need to be set. I'm just not sure why it bothers to null check before setting it.

Reply:
Are you asking for this to be changed to the following @davidfowl?

```c#
var headers = _values.Headers = new Dictionary<string, StringValues>(StringComparer.OrdinalIgnoreCase);
```

So `_values.Headers` is always set to an empty dictionary at the start of the middleware? @alefranz is correct that this is breaking. Here's an example of the code this would break. Not sure how realistic/common this is, but we probably want to be careful in a patch.

```c#

public void ConfigureServices(IServiceCollection services)
{
    services.AddHeaderPropagation(options =>
    {
        options.Headers.Add("User-Agent");
    });
})

// ...

public void Configure(IApplicationBuilder app, HeaderPropagationValues headerPropagationValues)
{
    app.Use((_, next) =>
    {
        // Always "propagate" a "Foo: Bar" header.
        headerPropagationValues.Headers = new Dictionary<string, StringValues>(StringComparer.OrdinalIgnoreCase)
        {
            { "Foo", "Bar" }
        };

        return next();
    });

    app.UseHeaderPropagation();

    app.Run(async context =>
    {
        context.Response.Headers["Content-Type"] = "text/plain";

        foreach (var entry in headerPropagationValues.Headers)
        {
            await context.Response.WriteAsync($"{entry.Key}: {entry.Value}\r\n");
        }
    });
}
```

Today, this would print "Foo: Bar" to the response in addition to the propagated User-Agent. After newing up the dictionary every time, it would only include the propogated User-Agent.


Unless you mean the following:

```c#
var headers = _values.Headers = _values.Headers != null ?
    new Dictionary<string, StringValues>(_values.Headers, StringComparer.OrdinalIgnoreCase) :
    new Dictionary<string, StringValues>(StringComparer.OrdinalIgnoreCase);
```

That makes some sense. It would allow my sample to use a static "Foo: Bar" dictionary instead of newing one up each time since it would have to worry about HeaderPropagationMiddleware mutating it. Technically is someone was relying on the mutation, that could be breaking too, but I think that's less likely to be an issue.